url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/mvd/MVDFile.java

```diff
@@ -0,0 +1,771 @@
+        {
+            e.printStackTrace();
+        }
+        finally
+        {
+           	if (rs != null) {
+                try {
+                    rs.close();
+                } catch (SQLException sqlEx) { } // ignore
+
+                rs = null;
+            }
+            if (stmt != null) {
+                try {
+                    stmt.close();
+                } catch (SQLException sqlEx) { } // ignore
+
+                stmt = null;
+            if (conn != null)
+            {
+                 }
+                try
+                {
+                    conn.close ();
+                }
+                catch (Exception e) { /* ignore close errors */ }
+            }
+        }
+	}
+	/**
+	 * Write to an actual file
+	 * @param dst the location of the file
+	 * @param str the string of data to write
+	 * @throws Exception
+	 */
+	private static void writeToFile( File dst, String str ) throws Exception
+	{
+		if ( dst.exists() )
+			dst.delete();
+		dst.createNewFile();
+		FileWriter fw = new FileWriter( dst );
+		fw.write( str );
+		fw.close();
+	}
+	/**
+	 * Save an MVD to a file
+	 * @param mvd the MVD to save
+	 * @param dst the file to save it to
+	 * @param rb database properties file
+	 * @param folderId id of the folder to contain it in
+	 * @throws Exception raised if an error occurred
+	 */
+	public static void externalise( MVD mvd, File dst, int folderId, 
+			Properties rb ) throws Exception
+	{
+		long start = System.nanoTime()/1000;
+        System.gc();
+        long startMem = Runtime.getRuntime().freeMemory();
+        int size = mvd.dataSize();
+		byte[] data = new byte[size];
+		int nBytes = mvd.serialise( data );
+		assert nBytes==size: "MVD shorter than predicted";
+		String str = Base64.encodeBytes( data, Base64.GZIP );
+		if ( rb == null )
+			writeToFile( dst, str );
+		else
+			writeToDatabase( dst.getName(), str, mvd.description, folderId, rb );
+        long end = System.nanoTime()/1000;
+        long endMem = Runtime.getRuntime().freeMemory();
+        System.out.println("internalise took "+(end-start)+" microseconds");
+        System.out.println("Memory used "+(startMem-endMem)+" bytes");
+        
+	}
+    /**
+	 * Convert an MVD to a string
+	 * @param mvd the MVD to convert
+	 * @param rb database properties file
+	 * @param folderId id of the folder to contain it in
+	 * @throws Exception raised if an error occurred
+	 */
+	public static String externalise( MVD mvd ) throws Exception
+	{
+        int size = mvd.dataSize();
+		byte[] data = new byte[size];
+		int nBytes = mvd.serialise( data );
+		assert nBytes==size: "MVD shorter than predicted";
+		return Base64.encodeBytes( data, Base64.GZIP );
+	}
+	/**
+	 * Read from an mvd file on disk
+	 * @param src the src mvd
+	 * @return a char array with the file contents
+	 * @throws Exception
+	 */
+	private static char[] readFromFile( File src ) throws Exception
+	{
+		char[] data = null;
+		FileReader fr = new FileReader( src );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fr*. The resource is referenced at line 274. The resource is closed at line 275. There are other execution paths that do not contain closure statements, for example, when an else condition at line 270 evaluates to true. Ensure *fr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).